docs/internals/annotations_resolving.md

```diff
@@ -0,0 +1,267 @@
+following logic is applied:
+
+1. Fetch the `__annotations__` key from the current base class' `__dict__`, if present. For `Base`, this will be
+   `{'f1': 'MyType'}`.
+2. Iterate over the `__annotations__` items and try to evaluate the annotation [^1] using a custom wrapper around
+   the built-in [`eval()`][eval] function. This function takes two `globals` and `locals` arguments:
+     - The current module's `__dict__` is naturally used as `globals`. For `Base`, this will be
+       `sys.modules['module1'].__dict__`.
+     - For the `locals` argument, Pydantic will try to resolve symbols in the following namespaces, sorted by highest priority:
+         - A namespace created on the fly, containing the current class name (`{cls.__name__: cls}`). This is done
+           in order to support recursive references.
+         - The locals of the current class (i.e. `cls.__dict__`). For `Model`, this will include `LocalType`.
+         - The parent namespace of the class, if different from the globals described above. This is the
+           [locals][frame.f_locals] of the frame where the class is being defined. For `Base`, because the class is being
+           defined in the module directly, this namespace won't be used as it will result in the globals being used again.
+           For `Model`, the parent namespace is the locals of the frame of `inner()`.
+3. If the annotation failed to evaluate, it is kept as is, so that the model can be rebuilt at a later stage. This will
+   be the case for `f5`.
+
+The following table lists the resolved type annotations for every field, once the `Model` class has been created:
+
+| Field name | Resolved annotation |
+|------------|---------------------|
+| `f1`       | [`int`][]           |
+| `f2`       | [`str`][]           |
+| `f3`       | [`bool`][]          |
+| `f4`       | [`bytes`][]         |
+| `f5`       | `'UnkownType'`      |
+
+### Limitations and backwards compatibility concerns
+
+While the namespace fetching logic is trying to be as accurate as possible, we still face some limitations:
+
+<div class="annotate" markdown>
+
+- The locals of the current class (`cls.__dict__`) may include irrelevant entries, most of them being dunder attributes.
+  This means that the following annotation: `f: '__doc__'` will successfully (and unexpectedly) be resolved.
+- When the `Model` class is being created inside a function, we keep a copy of the [locals][frame.f_locals] of the frame.
+  This copy only includes the symbols defined in the locals when `Model` is being defined, meaning `InnerType2` won't be included
+  (and will **not be** if doing a model rebuild at a later point!).
+- To avoid memory leaks, we use [weak references][weakref] to the locals of the function, meaning some forward references might
+  not resolve outside the function (1).
+- Locals of the function are only taken into account for Pydantic models, it does not apply to dataclasses, typed dictionaries
+  or named tuples.
```

Comment:
Can we make these sub-bullets of the above bullet given that they all refer to the function context?